_posts/2017-12-07-mixer-spof-myth.md

```diff
@@ -0,0 +1,108 @@
+
+## Architecture recap
+
+As shown in the diagram below, Mixer sits between the mesh components and the infrastructure backends that support it:
+
+<figure><img src="./img/mixer-spof-myth-2.svg" alt="Istio Topology" title="Istio Topology" />
+<figcaption>Istio Topology</figcaption></figure>
+
+The proxy logically calls Mixer before each request to perform precondition checks, and after each request to report telemetry. The proxy has local caching such that a relatively large percentage of precondition checks can be performed from cache. Additionally, the proxy buffers outgoing telemetry such that it only actually needs to call Mixer once for every several thousands requests. Whereas precondition checks are synchronous to request processing, telemetry reports can be done asynchronously with a fire-and-forget pattern.
+
+At a high level, Mixer provides:
+
+* **Backend Abstraction**. Mixer insulates the Istio components and services within the mesh from the implementation details of individual infrastructure backends.
+
+* **Intermediation**. Mixer allows operators to have fine-grained control over all interactions between the mesh and the infrastructure backends.
+
+However, even beyond these purely functional aspects, Mixer has other characteristics that provide the system with additional benefits.
+
+## Mixer: SLO booster
+
+Contrary to the claim that Mixer is a SPOF and can therefore lead to mesh outages, we believe it in fact improves the effective availability of a mesh. How can that be? There are three basic characteristics at play:
+
+* **Statelessness**. Mixer is stateless in that it doesn’t manage any persistent storage of its own.
+
+* **Hardening**. Mixer proper is designed to be a highly reliable component. The design intent is to achieve > 99.999% uptime for any individual Mixer instance.
+
+* **Caching and Buffering**. Mixer is designed to accumulate a large amount of transient ephemeral state.
+
+The sidecar proxies that sit next to each service instance in the mesh must necessarily be frugal in terms of memory consumption, which constrains the possible amount of local caching and buffering. Mixer, however, lives independently and can use considerably larger caches and output buffers. Mixer thus acts as a highly-scaled and highly-available second-level cache for the sidecars.
+
+Mixer’s expected availability is considerably higher than most infrastructure backends (those often have availability of perhaps 99.9%). Its local caches and buffers help mask infrastructure backend failures by being able to continue operating even when a backend has become unresponsive.
+
+## Mixer: Latency slasher
+
+As we explained above, the Istio sidecars generally have fairly effective first-level caching. They can serve the majority of their traffic from cache. Mixer provides a much greater shared pool of second-level cache, which helps Mixer contribute to a lower average per-request latency.
```

Comment:
per-request latency for the calls to the infrastructure backends ?

Reply:
Actually, it reduces the latency for anything incoming to the mesh. This is because the sidecars are synchronously performing precondition checks before allowing any request in. If we speed up our handling of precondition checks, the the proxy responds faster and thus the effective latency of the underlying service is improved.